File: java/medium/lengthOfLongestSubstring.java
/* This program takes a string s and finds the length of the longest
substring within it that does not have repeating characters
Rating: Medium
*/

class Solution {
    public int lengthOfLongestSubstring(String s) {

        String longest = "";

        if (s.length() > 50000) {
            System.out.println("Invalid string length");
            return 0;
        }

        for (int i = 0; i < s.length(); i++) {
            String temp = "";
            temp += s.charAt(i);
            int j = i+1;
            while ((j <= (s.length() - 1)) && (!temp.contains(String.valueOf(s.charAt(j))))) {
                temp += s.charAt(j);
                j++;
            }
                if (temp.length() > longest.length()) {
                    longest = temp;
                }
            }
            
        return longest.length();

        }

    }

File: python/medium/lengthOfLongestSubstring.py
# This program takes a string s and finds the length of the longest
# substring within it that does not have repeating characters
# Rating: Medium
# Performance: Runtime O(N) -> beats ~90%

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """

        seen = {}
        l = 0
        length = 0
        for r in range(len(s)):
            char = s[r]
            if char in seen and seen[char] >= l:
                l = seen[char] + 1
            else:
                length = max(length, r - l + 1)
            seen[char] = r
        
        return length